go-controller/pkg/clustermanager/healthcheck/controller.go

```diff
@@ -54,6 +54,15 @@ func transitionNodeState(old nodeState, health HealthState, on time.Time, timeou
 			new.health = UNREACHABLE
 			reason = fmt.Sprintf("not available since %s", old.available)
 		}
+	case UNAVAILABLE:
+		// only if we were available, otherwise services that chose to schedule
```

Comment:
not sure I quite unerstand the comment above. If we are moving the state to unavailable, why would we care about the previous state being unknown?

Reply:
UNKNOWN -> UNAVAILABLE -> AVAILABLE is a potentially expected transition  for a node as it starts up.

Other option I considered is to promote a certain amount of UNAVAILABLE time to UNREACHABLE but that also felt a bit arbitrary and unsure if it was something we would like to do here or leave it up to the consumers who decided to risk it and schedule services with UNKNOWN state.

I m sure right now if there was something else. Open to suggestions.

Reply:
And definitely I can have a better comment there.

Reply:
So you are saying to declare something unavailable, you want to make sure it was available some point before. Otherwise you dont know if the dataplane was ever setup correctly?

I think promoting unavailable to unreachable actually makes sense. If a node is rebooting, there will be some time while it terminates processes for shutdown where you might still be getting TCP responses, but the node cannot serve as egress. This does seem like a slippery slope though. I think maybe only by leveraging other status from kapi could we make the judgement call of "is this node not responding correctly because OVNK pods are being upraded/restarted". Maybe information like node object status, ovnk pod status.

It might be better to defer this unavailable functionality to another PR where we can discuss separately.

Reply:
>So you are saying to declare something unavailable, you want to make sure it was available some point before. Otherwise you dont know if the dataplane was ever setup correctly? 

Ah yeah, sorry, I read that wrong.

> This does seem like a slippery slope though.

So the idea of UNAVAILABLE is to avoid unscheduling services that can still be served while avoiding scheduling new services in the node. And it is a slippery slope but I think right now we are overly jealous with it and not handling correctly an expected and frequent scenario like an ovnkube rollout which is a costly price to pay to handle some other unexpected scenarios.

So I think we need to be sure that a node is dead rather than moving things around at the minimum blip which is causing a lot of unnecessary downtime to users. 
To the extent that I would just shift the approach now and try to adjust in the future rather than delay it much more while we come up with the perfect way of doing it.

> If a node is rebooting, there will be some time while it terminates processes for shutdown where you might still be getting TCP responses, but the node cannot serve as egress

We are getting TCP responses through the management port. I interpret that as the dataplane being available in general. So that anything that was already configured on the node should still work, we just cant configure new things. I think this is the incorrect interpretation in this case, even if there is a small time window where I would get TCP responses but the data plane would be dead for whatever reason.

> I think promoting unavailable to unreachable actually makes sense. 

What I meant is something like "If a node is UNAVAILABLE for 30s, lets set it as UNREACHABLE". Avoided it because deciding what those 30s actually need to be seemed arbitrary.

> Maybe information like node object status, ovnk pod status.

I couldn't find a good way of leveraging this. ovnk pod status doesn't tell us much more than what the actual serving healthcheck is telling us.  For node object status it doesn't tell us much more than a "connection refused" or timeout to the healthcheck is telling us, also we are not in control nod status relevant timeouts, but it might be useful in hypershift hosted clusters although I didn't want to target that scenario in this PR.

> It might be better to defer this unavailable functionality to another PR where we can discuss separately.

We can do that, but I don't want to do it at a cost of degrading the importance of solving it.